### ISSUE TITLE
Player: CSV Utils crash randomly

### ISSUE THREAD

**Issue by ckbaumann** [pupil-labs/pupil#1304]
iMotions export works fine most of the time, but sometimes crashes after pressing [e] for the same recording. This does not depend on the selected timerange. The file 'info.csv' that is opened here looks fine and does not change from non-crashing to crashing exports. I can provide a recording folder for reproducing the issue.

```
Traceback (most recent call last):
  File "/cluster/users/Chris/git/pupil/pupil_src/launchables/player.py", line 452, in player
    p.on_notify(n)
  File "/cluster/users/Chris/git/pupil/pupil_src/shared_modules/imotions_exporter.py", line 140, in on_notify
    self.export_data(notification['range'], notification['export_dir'])
  File "/cluster/users/Chris/git/pupil/pupil_src/shared_modules/imotions_exporter.py", line 163, in export_data
    rec_start = self.get_recording_start_date()
  File "/cluster/users/Chris/git/pupil/pupil_src/shared_modules/imotions_exporter.py", line 242, in get_recording_start_date
    rec_info = csv_utils.read_key_value_file(csvfile)
  File "/cluster/users/Chris/git/pupil/pupil_src/shared_modules/csv_utils.py", line 28, in read_key_value_file
    dialect = csv.Sniffer().sniff(first_line, delimiters=',\t')
  File "/usr/lib/python3.6/csv.py", line 188, in sniff
    raise Error("Could not determine delimiter")
_csv.Error: Could not determine delimiter
```

**Comment by Windsooon** [comment]
Hello, @ckbaumann It will be great if you can provide a recording folder so we can fix it.

**Comment by papr** [comment]
Hey @Windsooon 
Thank you very much for your interest in this issue! You can use this recording:
https://drive.google.com/drive/folders/1OugotQQHsrO42S0CXwvGAa0HDvZ_WChG?usp=sharing

**Comment by Windsooon** [comment]
@papr I just get involved in the project :D. I found the get_recording_start_date function already been deleted in commit a63db85. @ckbaumann would you mind updating the trackback message?

**Comment by ckbaumann** [comment]
@Windsooon I now got the following stack trace a few times. Its a really weird bug, it once didn't crash for like 20 exports and at other times it crashed immediately or after only 2-3 exports. Maybe it's related to the selected export range, but even if you keep the range the same it often crashes after a few exports.

```
Traceback (most recent call last):
  File "/cluster/users/Chris/git/pupil/pupil_src/launchables/player.py", line 454, in player
    p.on_notify(n)
  File "/cluster/users/Chris/git/pupil/pupil_src/shared_modules/video_exporter.py", line 153, in on_notify
    self._start_export(notification["range"], notification["export_dir"])
  File "/cluster/users/Chris/git/pupil/pupil_src/shared_modules/video_exporter.py", line 182, in _start_export
    self.export_data(export_range, export_dir)
  File "/cluster/users/Chris/git/pupil/pupil_src/shared_modules/eye_video_exporter.py", line 79, in export_data
    self._export_eye_video(export_range, export_dir, eye_id=0)
  File "/cluster/users/Chris/git/pupil/pupil_src/shared_modules/eye_video_exporter.py", line 72, in _export_eye_video
    process_frame=process_frame,
  File "/cluster/users/Chris/git/pupil/pupil_src/shared_modules/video_exporter.py", line 197, in add_export_job
    rec_start = self._get_recording_start_date()
  File "/cluster/users/Chris/git/pupil/pupil_src/shared_modules/video_exporter.py", line 235, in _get_recording_start_date
    rec_info = csv_utils.read_key_value_file(csvfile)
  File "/cluster/users/Chris/git/pupil/pupil_src/shared_modules/csv_utils.py", line 28, in read_key_value_file
    dialect = csv.Sniffer().sniff(first_line, delimiters=',\t')
  File "/usr/lib/python3.6/csv.py", line 188, in sniff
    raise Error("Could not determine delimiter")
_csv.Error: Could not determine delimiter
```

**Comment by Windsooon** [comment]
@ckbaumann, Usually this happens when the info.csv is empty, or we get an empty delimiter in [sniff function](https://github.com/python/cpython/blob/master/Lib/csv.py#L176). A quick fix will check the delimiter is empty or not then raise error message or move forward. Maybe you can upload the CSV file google drive and I will try to reproduce the bug.

**Comment by ckbaumann** [comment]
@Windsooon The problem is that the info.csv is not empty and with the same files sometimes it crashes and sometimes it doesn't. I think you could use the example above provided by @papr that also includes an info.csv. Best of luck, we think that this is a tricky one!

**Comment by Windsooon** [comment]
I tried all the three info.csv (include the one in 863-998 and 1054-1201 folder),  I can't reproduce the bug. Which info.csv are from crashing export?

**Comment by papr** [comment]
This issue is a tricky race condition!

Two things happen in parallel:

1.1. The world video exporter updates the recording in the background (because of legacy reasons) https://github.com/pupil-labs/pupil/blob/54d27c991f4a2093a808da94f1eefc8c24efb443/pupil_src/shared_modules/exporter.py#L58
1.2. This overwrites the `info.csv` file in a non-atomic fashion: https://github.com/pupil-labs/pupil/blob/master/pupil_src/shared_modules/player_methods.py#L268-L269
2. The iMotions Exporter loads the `info.csv` in the foreground to read the recording date.

---

Things that need to be done:
1. Remove the legacy recording update
2. Make sure that all non-incrementally written files use https://boltons.readthedocs.io/en/latest/fileutils.html#atomic-file-saving
3. Refactor `csv_utils` to take a path as argument instead of a file handle. This should remove a lot of duplicated `open()` code, e.g. in `player_methods`.

**Comment by Windsooon** [comment]
@papr Great! I also found we have a very big [try](https://github.com/pupil-labs/pupil/blob/54d27c991f4a2093a808da94f1eefc8c24efb443/pupil_src/shared_modules/exporter.py#L580) which make us hard to debug. I also found some code about **while True** may also cause race condition. However, the first thing we would be write more tests.

**Comment by papr** [comment]
This function is executed in the background via a `Task_Proxy`:
https://github.com/pupil-labs/pupil/blob/54d27c991f4a2093a808da94f1eefc8c24efb443/pupil_src/shared_modules/background_helper.py#L23-L25

Exceptions yielded in the background are raised in the foreground.

Can you elaborate on the race conditions in regard to the while-loop?

**Comment by Windsooon** [comment]
Yes, I found some of them may cause race conditions, after I finish the ls project I will create an issue/pr.

**Comment by Windsooon** [comment]
I think we can also use [share lock](https://docs.python.org/3/library/fcntl.html#fcntl.flock) to avoid race condition whenever we write a file.

**Comment by Windsooon** [comment]
IMO, we can replace the with open code to:

```
def write_csv_file(file_path, data):
    while True:
        csvfile = open(file_path, 'w', newline="", encoding="utf-8")
        try:
            # try to get an exclusive lock
            fcntl.flock(csvfile, fcntl.LOCK_EX | fcntl.LOCK_NB)
        except IOError:
            # try to get the lock 0.1 second later
            time.sleep(0.1)
        else:
            csv_utils.write_key_value_file(csvfile, data)
            fcntl.flock(csvfile, fcntl.LOCK_UN)
            csvfile.close()
            break
```

**Comment by papr** [comment]
The actual concurrency issue will be fixed with #1322 via 11b12901f9ecc222cd2e90b32b93863af7bd1903

I think that we should use the file lock mechanism anyway. I also think that we should apply this to all files that we write, not only csv files.

I created a new issue for this and assigned the v1.10 milestone: #1363

**Comment by papr** [comment]
Fixed via #1322
